### PR TITLE
Add class methods for plugins registration in Connections class

### LINKED ISSUE #142: expose `available_connections` via InitNornir
So users can specify extra connection types

### LINKED ISSUE #211: Augment Connections class
This class could contain available connections as class attributes and instance attributes would be existing connections. 
Here is my previous proposal, which should be tested / discussed more:
```
class HostConnections(Dict[str, ConnectionPlugin]):
    available: Dict[str, Type[ConnectionPlugin]] = {}

    @classmethod
    def register(cls, name, value):
        cls.available[name] = value

    def open(self, name: str, ...):
        connection = self.available[name]()
        connection.open()
        self[name] = connection

    def close(self, ...): ...

    def close_all(self):
        for connection in self:
            self.close(connection)

    def get_or_create(self, ...): ...


class Host(object):
    ...
    def __init__(self, ...):
        ...
        self.connections = HostConnections()

    def open_connection(self, name: str, ...):
        self.connections.open(name, ...)
    ...
    
# can be called from plugin code or user code
HostConnections.register('netmiko', Netmiko)
HostConnections.register('napalm', Napalm)

```

### PR DESCRIPTION
**Pull request by dmfigol**
This API can be used internally for built-in plugins and
externally by a developer of a custom connection plugin

Usage:
```
from nornir.core.connections import Connections, ConnectionPlugin

class MyConnectionPlugin(ConnectionPlugin):
    ...


Connections.register('my', MyConnectionPlugin)
plugin = Connections.get_plugin('my')  # used internally in open_connection
Connections.deregister('my')
```

Fixes #211 
Closes #142

### CODE REVIEW

**Comment by dmfigol** [comment]
Fix #211

**Comment by dmfigol** [comment]
Also I have two questions:
1. What is the proper place to register built-in plugins?
Currently, it is done in `nornir/plugins/connections/__init__.py`
This also means that `nornir.plugins.connections` should be imported somewhere in core.
It is currently in `nornir/core/__init__.py`:
`from nornir.plugins import connections  # noqa: W0611`
This import is required to execute registration code
2. Do we want to move all connection handling from `Host` class to `Connections` class?

**Comment by dbarrosop** [comment]
Gave it a quick glance and I think it looks great. I need to review it a bit more thoroughly still but looks quite good. Regarding your questions:

1. Maybe we can put it in a method `register_default_plugins()` so we import it in `nornir/core/__init__.py` and then call it in the global scope. In general, we should try to avoid hidden side effects so it's better if we make the registration explicit this way.
2. No, all that logic conceptually belongs to the `Host` itself as it's the placeholder for data so I think that should stay there.

Nice job :)

**Comment by dmfigol** [comment]
@dbarrosop I agree. Should this function be called with `__init__.py` or may be it should be done in `InitNornir`? (If latter we would also need to adjust tests a little)

File: nornir/plugins/connections/__init__.py [hunk 1]
```diff
@@ -1,16 +1,11 @@
...
+from nornir.core.connections import Connections
 
 
-available_connections: Dict[str, Type["ConnectionPlugin"]] = {
-    "napalm": Napalm,
-    "netmiko": Netmiko,
-    "paramiko": Paramiko,
-}
+def register_default_connection_plugins() -> None:
+    Connections.deregister_all()
```
[the remark this answers is not in the document]
Comment (dmfigol) [reply nornir/plugins/connections/__init__.py]: Original version was without `deregister_all`. The reason why I added it, to make sure that if you call `register_default_connections_plugins()` again (most likely accidentally), the end result is the same.

File: nornir/core/connections.py [hunk 2]
```diff
@@ -53,4 +57,63 @@ def close(self) -> NoReturn:
...
+                even if a connection plugin with the same name was already
+                registered. Default - False
+
+        Raises:
+            :obj:`nornir.core.exceptions.ConnectionPluginAlreadyRegistered`
+        """
+        if not force and name in cls.available:
+            raise ConnectionPluginAlreadyRegistered(
+                f"Connection {name!r} was already registered"
+            )
```
[the remark this answers is not in the document]
Comment (dmfigol) [reply nornir/core/connections.py]: I don't know the right answer. That's why I added `force=True` for flexibility

File: nornir/core/connections.py [hunk 3]
```diff
@@ -53,4 +57,63 @@ def close(self) -> NoReturn:
...
+        cls.available.pop(name)
+
+    @classmethod
+    def deregister_all(cls) -> None:
+        """Deregisters all registered connection plugins"""
+        cls.available = {}
+
+    @classmethod
+    def get_plugin(cls, name: str) -> Type[ConnectionPlugin]:
```
[the remark this answers is not in the document]
Comment (dmfigol) [reply nornir/core/connections.py]: I don't like `get_connection_type` but open to change the name if needed.
That said, `get_plugin` is called only internally in `open_connection` - a user does not need to call that.

File: nornir/plugins/connections/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ktbyers) [reply nornir/plugins/connections/__init__.py]: Except you would also remove any other drivers that had been added in-between the two calls.

File: nornir/core/connections.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ktbyers) [reply nornir/core/connections.py]: Yes, I think multiple calls with the same driver name and same class should do nothing and not raise an exception (this also relates that other deregister_all() operation in connections/__init__.py).

File: nornir/core/connections.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ktbyers) [reply nornir/core/connections.py]: Okay, how about `get_connection_plugin`?

File: nornir/plugins/connections/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dbarrosop) [reply nornir/plugins/connections/__init__.py]: I agree with kirk, we have two options, either we rename the method to `reset_connection_plugins` or we remove the `Connection.deregister_all`. I vote for the latter.

File: nornir/core/connections.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ktbyers) [reply nornir/core/connections.py]: Good point...I probably missed the context it would be invoked so I am fine with with `get_plugin`.

**Comment by dmfigol** [comment]
I think I addressed all of the comments, ready for another review.

**Comment by dbarrosop** [comment]
Ok, looks good to me. I will let kirk give it another look if he wants to before merging.

Nice job :)

**Comment by ktbyers** [comment]
Looks good to me.

**Comment by ktbyers** [comment]
@dmfigol You can go ahead and merge this.

**Pull request merged**